### DISCUSSION TITLE
auto fail mod/customization options

### DISCUSSION THREAD

**Discussion by iSimpFor** [ppy/osu#13120] [Ideas]
this mod will force a fail to occur if the requirements are not met

the customization options are:
accuracy(if below, map is failed)
hp(if below, map is failed) 
a total amount/percentage of any judgement rating(if amount/percentage of not-best ratings is higher than threshold, map is failed or if amount/percentage of best ratings is lower than threshold, map is failed)

all options are disabled by default
an option to make a check on breaks or at the end of the map

these options can be put into customization box without the mod(meaning that the customization button will always be available to access these options)

**Comment by doughn0** [comment]
Also, an option to fail if it is no longer possible to beat your highscore on the map with the current play.

**Reply by stevefan1999-personal** [reply]
Instead of beating your high score, let's say we need to beat our own PP record anyway. This makes analogy to the same parity provided by Maimai/Chunithm's track skip feature.

**Comment by enoslayd** [comment]
For mania (and maybe other modes as well), have at least x combo when finishing a beatmap (ex. 200 combo at the end to consider a pass)

**Comment by Joehuu** [comment]
> Also, an option to fail if it is no longer possible to beat your highscore on the map with the current play.

Idea already in https://github.com/ppy/osu/issues/6642.

**Comment by GSculerlor** [comment]
Looking into this one. Should it included on fun mod or can be included on same group with SD/PF? And also any idea about the name? Since I don't think "Auto Fail" is good name here

**Comment by bdach** [comment]
I'm not sure all of these proposed mechanics should be one customisable mod; seems like the mod would be doing too much to give it a single intelligible name or purpose. splitting per proposed fail condition (i.e. something like *Score Challenge*, *Accuracy Challenge*) might fare better. those then could be combined if need be, I suppose.

**Comment by iSimpFor** [comment]
maybe we should put these options into a new section called "challenge section" and also move sd/pf there

it's a section where you challenge yourself to do better after all so all those mods will fit

**Comment by peppy** [comment]
New section sounds fine.

**Comment by MitsukaiDokuro** [comment]
Would prefer this implemented as Sudden Death mod options instead of a new mod.

**Comment by ramjam18** [comment]
I suggest a dropdown menu in Mod Customization for SuddenDeath instead of seperate mods.

A SuddenDeath with the name "name1" with fail condition [a] and a SuddenDeath called "name2" with fail condition [b]...
They all do the same thing; **they cause the player to fail** if a certain condition is met. So imo, it would make more sense to classify them all as _one_ SuddenDeath, and instead give this SuddenDeath customization options.


As for options for SuddenDeath, I've thought of a condition: _cause the player to fail if they don't alternate_. In other words, **they cannot press the same key twice in a row**. Perhaps name this condition "**MustAlternate**"? Just a suggestion ofc.

Also, I **support the idea of opening a new Challenge Section**. SuddenDeath etc. didn't really make the map harder per se, so I think, too, that moving them from Difficulty Increase to this new section would make sense.

**Reply by bdach** [reply]
> I suggest a dropdown menu in Mod Customization for SuddenDeath instead of seperate mods.

separate mods would not allow you to combine multiple failure modes which you could naturally do if they were combined.

> As for options for SuddenDeath, I've thought of a condition: cause the player to fail if they don't alternate

https://github.com/ppy/osu/pull/9606

**Reply by ramjam18** [reply]
> separate mods would not allow you to combine multiple failure modes which you could naturally do if they were combined.

uuhhh, do you mean that seperating the mods (conditions) would allow for them to be combined? cause if yes, then yeah, i guess you're right...

> #9606

... goddamnit, i'm always so late

**Reply by MitsukaiDokuro** [reply]
> > I suggest a dropdown menu in Mod Customization for SuddenDeath instead of seperate mods.
> 
> separate mods would not allow you to combine multiple failure modes which you could naturally do if they were combined.

How about transforming the fail conditions from a radio button group to a checkbox group; that'll
solve the combination issue surely.

**Reply by ramjam18** [reply]
Agree. So long as the selected conditions don't conflict with each other, that might work.

**Comment by abraker95** [comment]
I believe this is basically the selective perfect mod request on forums with almost 2000 votes

https://osu.ppy.sh/community/forums/topics/141061

**Comment by old-account123** [comment]
This idea would work great with playlists! Setting specific clear conditions for beatmaps would make for some challenging playlists.

**Comment by ggliv** [comment]
I've made a barebones version of this mod proposal with customizable accuracy, health, and judgement thresholds for the Osu ruleset. You can check it out here: https://github.com/ggliv/osu/commit/8320ea8d91295c6f4dc75f1264c82652992001b4.

I'm going to continue working on it and make a PR once it's in a more complete state. Feel free to use what's there now.

(Please let me know if this should be in a separate discussion)

**Comment by ggliv** [comment]
I've gotten the mod to a point where it meets all the criteria of the original post except for having all options disabled by default. Having the non-best judgement count condition disabled by default is not immediately feasible. As an alternative, it may be appropriate to integrate the conditional failure checking into the perfect mod, since you can set all defaults with the assumption of requiring a perfect play.

Does anyone have any thoughts on this? "Perfect" by itself doesn't imply any failure checking customization... I'm not sure how the UX would work.

**Reply by bdach** [reply]
> Having the non-best judgement count condition disabled by default is not immediately feasible

first of all, why not?

**Reply by ggliv** [reply]
A non-best judgement threshold setting would be a slider from 0 to some number. To enable the judgement threshold, you would use a checkbox setting and use that as a condition in the failure logic to determine whether you need to check the judgement threshold. You would naturally want the judgement threshold slider to be 'greyed out'/nonadjustable when the checkbox is set to disable the judgement threshold. As far as I know, the only way do that would be to disable the slider's underlying bindable, which triggers exceptions and a crash when `ModSelectOverlay` tries to instantiate the mod.

**Reply by bdach** [reply]
i'm mostly interested in the "exception and a crash" part. what exception? what's the stack trace?

**Reply by ggliv** [reply]
```stacktrace
Unhandled exception. System.InvalidOperationException: Can not set value to "0" as bindable is disabled.
   at osu.Framework.Bindables.Bindable`1.set_Value(T value)
   at osu.Framework.Bindables.RangeConstrainedBindable`1.setValue(T value)
   at osu.Framework.Bindables.RangeConstrainedBindable`1.set_Value(T value)
   at osu.Framework.Bindables.BindableNumber`1.setValue(T value)
   at osu.Framework.Bindables.BindableNumber`1.set_Value(T value)
   at osu.Framework.Bindables.Bindable`1.BindTo(Bindable`1 them)
   at osu.Framework.Bindables.RangeConstrainedBindable`1.BindTo(Bindable`1 them)
   at osu.Framework.Bindables.BindableNumber`1.BindTo(Bindable`1 them)
   at osu.Framework.Bindables.Bindable`1.osu.Framework.Bindables.IBindable.BindTo(IBindable them)
   at osu.Game.Rulesets.Mods.Mod.CopyAdjustedSetting(IBindable target, Object source) in /home/ggliv/Desktop/osu/osu.Game/Rulesets/Mods/Mod.cs:line 171
   at osu.Game.Rulesets.Mods.Mod.CopyFrom(Mod source) in /home/ggliv/Desktop/osu/osu.Game/Rulesets/Mods/Mod.cs:line 153
   at osu.Game.Rulesets.Mods.Mod.DeepClone() in /home/ggliv/Desktop/osu/osu.Game/Rulesets/Mods/Mod.cs:line 135
   at osu.Game.Overlays.Mods.ModSelectOverlay.<>c.<createLocalMods>b__35_0(Mod mod) in /home/ggliv/Desktop/osu/osu.Game/Overlays/Mods/ModSelectOverlay.cs:line 288
   at System.Linq.Enumerable.SelectEnumerableIterator`2.ToArray()
   at osu.Game.Overlays.Mods.ModSelectOverlay.createLocalMods() in /home/ggliv/Desktop/osu/osu.Game/Overlays/Mods/ModSelectOverlay.cs:line 287
   at osu.Game.Overlays.Mods.ModSelectOverlay.<LoadComplete>b__31_0(ValueChangedEvent`1 _) in /home/ggliv/Desktop/osu/osu.Game/Overlays/Mods/ModSelectOverlay.cs:line 218
   at osu.Game.Overlays.Mods.ModSelectOverlay.LoadComplete() in /home/ggliv/Desktop/osu/osu.Game/Overlays/Mods/ModSelectOverlay.cs:line 218
   at osu.Framework.Graphics.Drawable.loadComplete()
   at osu.Framework.Graphics.Drawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Platform.GameHost.UpdateFrame()
   at osu.Framework.Threading.GameThread.processFrame()
```

**Reply by bdach** [reply]
this is not exactly an unsolvable problem. `CopyFrom` could well enable the target bindable for a second in order to copy across the value, and then copy across the disabled state from the source bindable.

**Reply by frenzibyte** [reply]
It's a known one (https://github.com/ppy/osu-framework/issues/3218), enabling target bindable in a similar manner to [this existing case](https://github.com/ppy/osu/blob/76c63f1d0a0a2bda6eb56642c0f417ba8656297c/osu.Game/Screens/Select/SongSelect.cs#L605-L607). I'm still unsure this will work as you may be greeted with a `SliderBar` crash due to not handling disabled bindables well on initial load.

**Reply by ggliv** [reply]
Made a rough solution that seems to be working: https://github.com/ggliv/osu/commit/a2fb26298e7e630e5c307f001d7db6f14e030925. I haven't included any tests but I can't immediately think of anything that this would cause issues for. I'm hesitant to include it in a potential PR since it messes with `SettingSourceAttribute`, though.

**Reply by frenzibyte** [reply]
I honestly don't know about that. I would much rather have the setting always enabled with default = 0 and call it a day.

**Reply by ggliv** [reply]
Is that preference a result of the restrictions on disabled bindables or is it what you would want any implementation of this mod to default to?

**Reply by frenzibyte** [reply]
Actually nevermind that, I suppose this mod is supposed to allow non-perfect judgements by default. Unsure of a better solution to that and would suggest waiting for a framework-side fix instead (will see if I can come up with something there).

**Reply by ggliv** [reply]
Thanks for giving it a look. I'm fine with waiting a bit for a framework change. Drop a PR link if/when you get to it.

**Comment by ggliv** [comment]
As an update to people following this, the mod is now completely implemented according to the original request. You can check it out yourself in [this branch](https://github.com/ggliv/osu/tree/mod-clearcondition). I'm holding off on creating a pull request for a little while since setting the judgement condition to be disabled by default requires some not-the-cleanest type checking/casting in outside functions, but it's completely usable if you just want to try it out.

Here's a video of the mod as it is now. Please let me know if you have any ideas/feedback/input.

https://user-images.githubusercontent.com/47010459/169391607-831c2cde-1734-4593-9548-b655e9d43722.mp4

**Reply by bdach** [reply]
very mixed feelings on the general UX of this if I'm honest. feels like everything and the kitchen sink was thrown into a single mod for no particular rhyme or reason. i shudder to even see the code given how many interdependent behaviour-altering option the mod has.

it probably needs to be split off into its constituent parts, i'd say, but i guess it's just a personal opinion.

**Reply by ggliv** [reply]
I personally think it makes at least a decent amount of sense for them to be combined, but I can definitely understand your viewpoint. The condition checking code isn't that bad, it just increments for judgements and accuracy, checks intervals, and returns a final boolean based on value/threshold comparisons. Maintainable from my vantage point, at least.

If the mod were to be split up into accuracy, health, and judgement portions, would you want to see them put into their own category as per peppy (https://github.com/ppy/osu/discussions/13120#discussioncomment-791594)? Should PF and SD go there too?

**Reply by iSimpFor** [reply]
i think there should be a yellow tint if you're close to any failing threshold and a red tint if you're under any failing threshold(if you're still alive), displayed like you're close to losing all hp in regular play

there could also be a leniency(maybe 10, 15, or 20 notes) before the mod actually starts checking clear conditions to give the player a bit of breathing room since things like first-note goods can happen

i think PF and SD can go in the new section since they fit the idea of "challenging yourself to do better/meet a threshold"

additionally, for the maximum amount of imperfect judgements, i think it should be a percentage(or a option to choose between amount of notes or percentage) since beatmaps can vary wildly in amount of notes

**Reply by Moltanicaa** [reply]
I can suggest moving the existing SD/PF into the subcategories of this mod so that they appear as an icon when properly configured.  For then it turns out that we have as many as 3 mods with similar functionality.

**Reply by GSculerlor** [reply]
I tried to work on this before (health one, particularly). I haven't checked your changes yet but probably you want to take a look at some consideration for draining health constraints here https://github.com/ppy/osu/pull/9584#issuecomment-659985472 (can be dismissed if you are already considering that concern on your changes)

**Reply by ggliv** [reply]
Separate section works well enough. @bdach, thoughts?

https://user-images.githubusercontent.com/47010459/169667223-cef3392f-8952-4a12-ae62-f04a54c663c9.mp4

**Reply by bdach** [reply]
seems much better to me at least, not sure what the others will think.

**Reply by Walavouchey** [reply]
looks great! i'd suggest the wording of "Checking: `Continuous` / `At breaks` / `At the end`"

i'm having a hard time understanding what "imperfect proportion" means though. imo it should let you specify a maximum amount of each judgement (by inputting a number), or at least i'd find that very useful as a player

**Reply by ggliv** [reply]
~~I don't think there's a 'roller' or some other config option with an unbounded top end, I'd probably  have to have a string input and make sure it's a number before the beatmap starts... not the cleanest but it'd work. I'll give a crack at it when I'm back at my computer.~~
Nevermind, that is precisely what `SettingsNumberBox` is.

**Reply by smoogipoo** [reply]
Afaik we weren't intending to ever expand on mod types. Would need @peppy to confirm.

**Reply by bdach** [reply]
[This comment](https://github.com/ppy/osu/discussions/13120#discussioncomment-791594) happened earlier, unless something has changed since.

**Reply by peppy** [reply]
I'm okay with new types being added as long as we can envisage more of the same mods being created going forward. And as long as they can be well defined (ie. with one sentence, we have unanimous agreement over which mods should exist in that section, or be moved there).

It should be done in a separate PR though, and after discussion, since some of the other existing mods would likely be moved depending on how the category is defined.

**Comment by ggliv** [comment]
https://github.com/ppy/osu/pull/18370
